### ISSUE TITLE
Serializer produces wrong association key format for polymorphic has-many relations

### ISSUE THREAD

**Issue by kaeufl** [samselikoff/ember-cli-mirage#1274]
There are two issues related to how polymorphic has-many relations are handled by the base Serializer:

Given the `user` model
```javascript
Model.extend({
    things: hasMany({ polymorphic: true })
})
```
1. If `serializeIds` is set to `'always'`,
the expected json (in accordance with what ember-data expects, as far as I know) would be 
```json
{
  "user": {
    "id": "1",
    "thingIds": [{
        "id": "1",
        "type": "picture"
      }]
  }
}
```

while the actual result is
```json
{
  "user": {
    "id": "1",
    "thingIds": ["1"]
  }
}
```
That is, the type information is missing from the polymorphic id.

2. If on the other hand `serializeIds` is set to `'included'`, and the 'things' relation is requested to be included in the payload, then to my understanding the expected json should be
```json
{
  "pictures": [
    {
      "id": "1",
      "title": "Lorem ipsum"
    }
  ],
  "user": {
    "id": "1",
    "name": "Ned",
    "thingIds":  [
          { "id": "1", "type": "picture" }
     ]
  }
}
```

while the actual result is
```json
{
  "pictures": [
    {
      "id": "1",
      "title": "Lorem ipsum"
    }
  ],
  "user": {
    "id": "1",
    "things": [
          { "id": "1", "type": "picture" }
     ]
  }
}
```

That is, the key is called "things" rather than "thingIds" in the serialization of the user model, although this time the type info is correctly reflected.

These issues can easily be worked around by overriding the serialize method in a custom serializer, but nevertheless, I feel that the current default behaviour is somewhat odd.

**Comment by geekygrappler** [comment]
Where are you getting the expected responses from? The PR looks OK, but I couldn't find what ED would expect here?

**Comment by geekygrappler** [comment]
Could you show me the setup for how you'd use `serializeIds` in an app because I can't get it to work

**Comment by kaeufl** [comment]
1. Regarding the format of the polymorphic id:
This is documented in the [guides](https://guides.emberjs.com/v3.0.0/models/relationships/#toc_polymorphism). See the "payment_methods" example. This not only holds true for the `JSONAPISerializer`, but also for the `RESTSerializer` (see https://github.com/emberjs/data/blob/252ffd7606a44a5c257bfb042cd7c7f51ae192d0/addon/serializers/rest.js#L779) and `JSONSerializer` (see https://github.com/emberjs/data/blob/252ffd7606a44a5c257bfb042cd7c7f51ae192d0/addon/serializers/json.js#L607), unfortunately less well documented.
2. Regarding the relationship key, i.e. `"things"` vs `"thingIds"`
On second thought, I'm actually not so sure anymore about what ED expects here by default. I forgot that we have custom behaviour in our app and I was thinking of [this example](http://www.ember-cli-mirage.com/docs/v0.4.x/serializers/#keyforrelationshipidsrelationship) in the mirage docs that seemed to suggest sth similar.
However, as far as I understand, still the [old code](https://github.com/samselikoff/ember-cli-mirage/blob/49ca35cec7c9c7ee4c6a9c2cd2ed26a76e16174f/addon/serializer.js#L291) made a mistake by calling the `keyForRelationship` hook, rather than the `keyForRelationshipIds` hook. If I read the docs correctly, the former is meant to define how the keys for any sideloaded collections should be formatted, whereas the latter is used to determine the format of the relationship keys themselves. See also http://www.ember-cli-mirage.com/docs/v0.4.x/serializers/#keyforrelationshiprelationship and http://www.ember-cli-mirage.com/docs/v0.4.x/serializers/#keyforrelationshipidsrelationship 

3. `serializeIds`
The `serializeIds` property can be set on a custom (mirage) serializer in order to specify under which circumstances relationship keys are added to the payload, like so:
```javascript
// serializers/my-custom-serializer.js
import Serializer from 'ember-cli-mirage';

export default Serializer.extend({
    serializeIds: 'always'
})
```
It defaults to `"included"`, i.e. only those keys are added for which data is sideloaded along with the main resource. This is documented here: http://www.ember-cli-mirage.com/docs/v0.4.x/serializers/#serializeids
If it helps, I can also try to provide a twiddle.

**Comment by geekygrappler** [comment]
TLDR: twiddle would be excellent please 👍

Hmmm so I wanted to test this out to understand what the behaviour should be. I made a simple app with some polymorphic relationships and tried to include the relationship Ids so I would see what ED wanted/expected. Here's the repo: https://github.com/geekygrappler/mirage-example.

My issue was that this doesn't ever include ids of relationships, if the relationship is requested (via `includes` in the request) then you get the info via relationships key as I'd expect from the guides.

```
{
  data: {
    attributes: {
      name: "Ned"
    }
    id: "1"
    relationships: {
      things: {
        data: [
          {
            id: "1",
            type: "pictures"
          }
        ]
      }
    }
    type: "users"
  }
  included: [ ... ]
}
```

After some digging I found out that with JSONAPI serializer the `serializeIds` key will always be ignored (which makes sense because I don't think serializing the IDs is part of the spec).

`_maybeAddAssociationIds()` is responsible for adding IDs: https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/serializer.js#L271, but with JSONAPI serializer the method that will eventually call this method, `getHashForPrimaryResource()` https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/serializer.js#L70, is replaced here: https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/serializers/json-api-serializer.js#L25

And this method will never check for serializing ids.

So I hope that makes sense. I'm wondering what serializer you're using in Mirage so I can get it to return a response with some serialized relationship IDs and see what ED is expecting, although I guess that also depends on what adapter you're using.

Basically I'm not sure what ED will expect with serailized Ids, the documentation is scant/non-existent, so I want to get a demo app working and see what keys/format ED does and doesn't like for this.

**Comment by kaeufl** [comment]
Yes, you are absolutely right. This seems to only be an issue when using the `RestSerializer` (and `ActiveModelSerializer` for that matter).

Since I was having some trouble getting the mirage twiddle to work, I took the liberty to fork your `mirage-example` app and make the necessary modifications in order to demonstrate the issue:
https://github.com/kaeufl/mirage-example/tree/demonstration-of-issue-%231274
(see the `demonstration-of-issue-#1274` branch and this [comment](https://github.com/kaeufl/mirage-example/blob/eac4c2b73f1848fa215cb2bed212d74153e4e801/mirage/serializers/user.js#L4) in the user serializer)

I also realized that the second issue in the bug report above, only reveals itself, when customizing the keyForRelationshipIds method, since the default implementations of both keyForRelationship and keyForRelationshipIds in the RestSerializer produce the same result.

**Comment by geekygrappler** [comment]
OK @kaeufl I just looked at the first problem, could you take a look at my [PR](https://github.com/samselikoff/ember-cli-mirage/pull/1282) and tell me if it's correct and solves the problem you're experiencing?

**Comment by samselikoff** [comment]
@kaeufl I'm going to start work on this soon, I agree the default behavior is strange.

Are you still around to help me with testing?

**Comment by kaeufl** [comment]
@samselikoff Sure! Let me know how I can be of help.
